crates/forge_app/src/tool_registry.rs

```diff
@@ -154,43 +160,99 @@ impl<S: Services> ToolRegistry<S> {
             }
         }
     }
-    #[allow(dead_code)]
-    pub async fn call(&self, input: ToolCallFull, context: &mut ToolCallContext) -> ToolResult {
-        let Ok(tool_input) = serde_json::from_value::<Tools>(input.arguments) else {
-            return ToolResult::new(input.name)
-                .failure(anyhow::anyhow!("Failed to parse tool input arguments"));
-        };
-
-        let result = self.call_internal(tool_input.clone(), context).await;
-        match result {
-            Ok(out) => {
-                let Ok(truncation_path) = out.to_create_temp(self.services.as_ref()).await else {
-                    return ToolResult::new(input.name)
-                        .failure(anyhow::anyhow!("Failed to truncate output"));
-                };
+    async fn call_forge_tool(
+        &self,
+        input: ToolCallFull,
+        context: &mut ToolCallContext,
+    ) -> anyhow::Result<ToolOutput> {
+        let tool_input =
+            serde_json::from_value::<Tools>(input.arguments).map_err(Error::ToolCallArgument)?;
 
-                let env = self.services.environment_service().get_environment();
+        let out = self.call_internal(tool_input.clone(), context).await?;
+        let truncation_path = out.to_create_temp(self.services.as_ref()).await?;
+        let env = self.services.environment_service().get_environment();
 
-                out.to_tool_result(input.name, Some(tool_input), truncation_path, &env)
-            }
-            Err(err) => ToolResult::new(input.name).failure(err),
+        out.into_tool_output(Some(tool_input), truncation_path, &env)
+    }
+
+    async fn call_mcp_tool(
+        &self,
+        input: ToolCallFull,
+        context: &mut ToolCallContext,
+        tool: Arc<Tool>,
+    ) -> anyhow::Result<ToolOutput> {
+        let output = tool.executable.call(context, input.arguments).await;
+        if let Err(error) = &output {
+            tracing::warn!(cause = ?error, tool = %input.name, "Tool Call Failure");
         }
+        output
+    }
+
+    async fn call_with_timeout<F, Fut>(
+        &self,
+        tool_name: &ToolName,
+        future: F,
+    ) -> anyhow::Result<ToolOutput>
+    where
+        F: FnOnce() -> Fut,
+        Fut: std::future::Future<Output = anyhow::Result<ToolOutput>>,
+    {
+        timeout(TOOL_CALL_TIMEOUT, future())
+            .await
+            .context(Error::ToolCallTimeout {
+                timeout: TOOL_CALL_TIMEOUT.as_secs() / 60,
+                tool_name: tool_name.clone(),
+            })?
     }
+
+    async fn call_inner(
+        &self,
+        input: ToolCallFull,
+        context: &mut ToolCallContext,
+    ) -> anyhow::Result<ToolOutput> {
+        tracing::info!(tool_name = %input.name, arguments = %input.arguments, "Executing tool call");
+        let tool_name = input.name.clone();
+
+        // First, try to call a Forge tool
+        if Tools::iter().any(|v| v.to_string().eq(input.name.as_str())) {
```

Comment:
```suggestion
        if Tools::contains(input.name) {
```
Should just check `contains`.